File: adhan/src/main/java/com/batoulapps/adhan/SunnahTimes.java
package com.batoulapps.adhan;

import com.batoulapps.adhan.data.CalendarUtil;
import com.batoulapps.adhan.data.DateComponents;

import java.util.Calendar;
import java.util.Date;

public class SunnahTimes {
    /* The midpoint between Maghrib and Fajr */
    public final Date middleOfTheNight;

    /* The beginning of the last third of the period between Maghrib and Fajr,
     a recommended time to perform Qiyam */
    public final Date lastThirdOfTheNight;

    public SunnahTimes(PrayerTimes prayerTimes) {
        final Date currentPrayerTimesDate = CalendarUtil.resolveTime(prayerTimes.dateComponents);
        final Date tomorrowPrayerTimesDate = CalendarUtil.add(currentPrayerTimesDate, 1, Calendar.DATE);
        final PrayerTimes tomorrowPrayerTimes =
                new PrayerTimes(prayerTimes.coordinates,
                        DateComponents.fromUTC(tomorrowPrayerTimesDate),
                        prayerTimes.calculationParameters);

        final int nightDurationInSeconds =
                (int) ((tomorrowPrayerTimes.fajr.getTime() - prayerTimes.maghrib.getTime()) / 1000);
        middleOfTheNight = CalendarUtil.roundedMinute(
                CalendarUtil.add(prayerTimes.maghrib, (int) (nightDurationInSeconds / 2.0), Calendar.SECOND));
        lastThirdOfTheNight =
                CalendarUtil.roundedMinute(
                        CalendarUtil.add(prayerTimes.maghrib,
                                (int) (nightDurationInSeconds * (2.0 / 3.0)),
                                Calendar.SECOND));
    }
}

File: adhan/src/commonMain/kotlin/com/batoulapps/adhan/SunnahTimes.kt
package com.batoulapps.adhan

import com.batoulapps.adhan.data.CalendarUtil
import com.batoulapps.adhan.data.CalendarUtil.add
import com.batoulapps.adhan.data.CalendarUtil.roundedMinute
import com.batoulapps.adhan.data.DateComponents
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.TimeZone.Companion
import kotlinx.datetime.toInstant

class SunnahTimes(prayerTimes: PrayerTimes) {
  /* The midpoint between Maghrib and Fajr */
  val middleOfTheNight: LocalDateTime

  /* The beginning of the last third of the period between Maghrib and Fajr,
     a recommended time to perform Qiyam */
  val lastThirdOfTheNight: LocalDateTime

  init {
    val currentPrayerTimesDate = CalendarUtil.resolveTime(prayerTimes.dateComponents)
    val tomorrowPrayerTimesDate = add(currentPrayerTimesDate, 1, DateTimeUnit.DAY)
    val tomorrowPrayerTimes = prayerTimes.copy(dateComponents = DateComponents.fromLocalDateTime(tomorrowPrayerTimesDate))

    val nightDurationInSeconds =
      (tomorrowPrayerTimes.fajr.toInstant(TimeZone.UTC).toEpochMilliseconds() -
          prayerTimes.maghrib.toInstant(Companion.UTC).toEpochMilliseconds()) / 1000
    middleOfTheNight = roundedMinute(
      add(prayerTimes.maghrib, (nightDurationInSeconds / 2.0).toInt(), DateTimeUnit.SECOND)
    )
    lastThirdOfTheNight = roundedMinute(
      add(
        prayerTimes.maghrib,
        (nightDurationInSeconds * (2.0 / 3.0)).toInt(),
        DateTimeUnit.SECOND
      )
    )
  }
}
